rfcs/56-better-reflect.md

```diff
@@ -0,0 +1,261 @@
+    type_info: TypeInfo,
+}
+impl TypeRegistration {
+  pub fn constructor(&self, proxy: ReflectProxy) ->  Result<Box<dyn Reflect>, ConstructorError> {
+    (self.constructor)(proxy)
+  }
+  // ...
+}
+```
+
+`constructor` would simply construct the `T` and wrap it into a `Box` before
+returning it.
+
+`constructor` can be derived similarly to `TypeInfo`. It will however
+require additional consuming methods on `Dynamic***` structs in order to be
+able to move their `Box<dyn Reflect>` fields into the constructed data
+structure.
+
+## Drawbacks
+
+- Users cannot define their own proxy type, they must use the `Dynamic***`
+  types.
+- `Dynamic***` do not implement the `Reflect` subtraits anymore.
+- There is some added complexity (although it fixes what I believe to be a
+  major limitation of the `bevy_reflect` API)
+- `Reflect::apply` now only accepts a `ReflectProxy` (see next section)
+
+## Rationale and alternatives
+
+- Divide `Reflect` in two distinct traits, one with the underlying-sensible
+  method, and one with the other methods. (see next points)
+- An earlier version of this RFC proposed `ReflectProxy` as a **trait**,
+  however, I reformulated it as an enum when considering how to implement it.
+  (How would you define the `TypeRegistration::constructor` field if passed a
+  `Box<dyn ReflectProxy>`? You'd need something like `ReflectProxyRef`?)
+
+I believe this change is needed, since `Reflect` duplicity is a large footgun.
+
+I don't believe that having `Dynamic***` implement `Reflect` makes sense, I
+think it introduces much more issues than it solves. The only use-case for
+`Dynamic***` being `Reflect` are the `reflect_mut`, `reflect_ref` and `apply`
+methods. However, those could as well be implemented outside of the `Reflect`
+trait, as mentioned in the "Future possibilities" section.
+
+## Unresolved questions
+
+- How does this interact with `#[reflect(ignore)]`?
+- Is a `trait ReflectProxy` more desirable than an `enum` and how to implement
+  it?
+- Should `Dynamic***` inner values be `Box<ReflectProxy>` or `Box<dyn Reflect>`?
+  With `Box<dyn Reflect>`:
+  * You must have a `TypeRegistry` and you must call `sidecast` both when
+    _building_ the `Dynamic***` and when converting from `ReflectProxy` to
+    `Reflect` (note that you most often already need the `TypeRegistry` when
+    building the `Dynamic***` if you want to access `TypeInfo`)
+  * The heavy lifting is mostly done when constructing the `ReflectProxy`,
+    `sidecast` is only a shallow conversion, calling `downcast` on the top
+    level data structure fields.
+  * There is less change to do if we keep it as it is today.
+  * With `Box<ReflectProxy>`, the opposite is true.
+- Should we try to implement the reflect subtraits for the `Dynamic***`?
+  (this would require removing the `Subtrait: Reflect` bounds)
+
+## Future possibilities
+
+We might be able to move `reflect_ref` and `reflect_mut` out of `Reflect` into
```

Comment:
Do we need to move these methods out?

Reply:
I think it's important to have some sort of replacement method on `Dynamic***` that does what `reflect_ref` does. Since it's not `Reflect` anymore, we'd lose those methods. Moving the `reflect_ref` to a trait that is both implemented by `dyn Reflect` (does rust even allow that???) and `ReflectProxy` would solve that.

Reply:
Yeah I see that point. I’m just hesitant on having `Reflect` be unique but not doing the same for its subtraits. Again, calling `as_reflect` on a `dyn Struct` should just work, but can't if the underlying type is a `DynamicStruct` according to this RFC.

Reply:
What would the implementation look like to make that work, though?